utils/node_local_test.py

```diff
@@ -2385,77 +2420,204 @@ def test_cont_clone(self):
                                       output = trace,
                                       test_class='test',
                                       duration = duration)
                 raise
             duration = time.time() - start
-            print('rc from {} is {}'.format(fn, rc))
-            print('Took {:.1f} seconds'.format(duration))
-            conf.wf.add_test_case(pt.test_name,
+            out_wrapper.sprint('Test {} took {:.1f} seconds'.format(function, duration))
+            conf.wf.add_test_case(ptl.test_name,
+                                  stdout = out_wrapper.get_thread_output(),
+                                  stderr = err_wrapper.get_thread_err(),
                                   test_class='test',
                                   duration = duration)
-            if not pt.needs_more:
+            if not ptl.needs_more:
                 break
-            pt.call_index = pt.call_index + 1
+            ptl.call_index = ptl.call_index + 1
 
     server.get_test_pool()
     pool = server.test_pool
 
-    pt = posix_tests(server, conf, pool=pool)
+    out_wrapper = nlt_stdout_wrapper()
+    err_wrapper = nlt_stderr_wrapper()
+
+    pto = posix_tests(server, conf, pool=pool)
     if test:
         fn = 'test_{}'.format(test)
-        obj = getattr(pt, fn)
+        obj = getattr(pto, fn)
 
-        _run_test()
+        _run_test(ptl=pto, test_cb=obj, function=fn)
     else:
 
-        for fn in sorted(dir(pt)):
+        threads = []
+
+        slow_tests = ['test_readdir_25', 'test_uns_basic', 'test_daos_fs_tool']
+
+        tests = sorted(dir(pto))
+        for stest in slow_tests:
+            tests.remove(stest)
+
+        all_tests = list(slow_tests)
+        all_tests.extend(tests)
+
+        for fn in all_tests:
             if not fn.startswith('test'):
                 continue
-            obj = getattr(pt, fn)
+
+            ptl = posix_tests(server, conf, pool=pool)
+            obj = getattr(ptl, fn)
             if not callable(obj):
                 continue
-            _run_test()
 
-    return pt.fatal_errors
+            thread = threading.Thread(None,
+                                      target=_run_test,
+                                      name='test {}'.format(fn),
+                                      kwargs={'ptl': ptl, 'test_cb': obj, 'function': fn},
+                                      daemon=True)
+            thread.start()
+            threads.append({'thread': thread, 'ptl': ptl})
+
+            # Limit the number of concurrent tests, but poll all active threads so there's no
+            # expectation for them to complete in order.  At the minute we only have a handlful of
+            # long-running tests which dominate the time, so whilst a higher value here would
+            # work there's no benefit in rushing to finish the quicker tests.  The long-running
+            # tests are started first.
+            while len(threads) > 5:
```

Comment:
Could a ThreadPool be used here?
https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.ThreadPool

Reply:
It could, but I'm not sure it's worth it.  The code as is uses the "deamon" flag, that does different things in the stdout handler based on that which would need re-writing so I don't think it's worth re-doing this.

Reply:
Fair enough :)